### ISSUE TITLE
Enable customers to use gitops tooling for apps (with Flux CD)

### ISSUE THREAD

**Issue by cokiengchiara** [giantswarm/roadmap#279]
**Business value**: Increase customer usage of app platform because it's increasing popularity in community and customers have been asking for it.

**User value:** Users can get the benefits of gitops in managing apps. There are many, incl. two important ones: (1) change approval and (2) having app state (versions in which clusters) and change control (when did that change) in git, rather than in customer's own pipeline (which they have to manage).

Ross: Yes I think we should write docs and maybe create an example repo with some of our apps to get up and running quickly

## Use cases:

**1/ Be able to configure managed apps upfront (before cluster creation) https://github.com/giantswarm/roadmap/issues/197**

**2/ Rollback changes if an upgrade has problems**

A gitops workflow can help make rollbacks easier. To provide greater visibility of changes and reverting those changes.
See https://github.com/giantswarm/roadmap/issues/337 for adding rollback to App Platform if sanity checks fail.

**3/ What Marcel wrote in RFC**
https://github.com/giantswarm/rfc/pull/2
(I still need to read)

**4/ Increase transparency and even alert on drifts between current and desired version.**

---

**JTBD** - a, b, o

Criteria for success:

- How many different places and repos do I have to copy and paste and make pull requests to, to configure an app with cluster-specific changes and do this promotion? (Best is ‘mono repo’.)
- How much time do I have to spend sorting our merge conflict? (The less, the better)
- How quickly can a new teammate understand how it works? (The quicker, the better)
- How much of this seems like ‘magic’ to an outsider? (The more transparent and easier it is to understand, the better)

Jobs:

- Share a global config that works across environments, but be able to override them for specific clusters. Avoid configuration duplication.
- Configure and test that an upgraded app works in playground env
- Configure and test that an upgraded app works in dev
- Configure and test that an upgraded app works in prod
- Deploy the upgraded app to all clusters

There are many others.

In other words, “How we do ‘promotion’ - testing that an upgraded app works in playground, then dev, then prod, before deploying everywhere

**Comment by cokiengchiara** [comment]
@giantswarm/team-batman adding this as candidate to next cycle

**Comment by rossf7** [comment]
Let's see if we can address https://github.com/giantswarm/roadmap/issues/197 as part of this.

So customers can configure apps before cluster creation.

**Comment by cokiengchiara** [comment]
@puja108 made this comment in [Jour Fixe slides](https://docs.google.com/presentation/d/1KgIdLLfZXR13MQ3xTa1OUCbtWS-Wgb6uIWWa7pMuTVE/edit#slide=id.gc68faa98da_1_0). Valuable as an MVP, if we can do it.

As we talked about this a bit on Slack, it might be interesting to include thoughts about how maybe gitops (with argo or flux) might be a good first MVP towards having basically  appsets (and in.a way other future stories like update functionality)

**Comment by rossf7** [comment]
> As we talked about this a bit on Slack, it might be interesting to include thoughts about how maybe gitops (with argo or flux) might be a good first MVP towards having basically appsets 

@cokiengchiara @puja108 @piontec Biscuit are looking at gitops tooling (Flux currently more likely than Argo) for replacing draughtsman. I think we should also write docs for how customers can use gitops to manage their app CRs. 

So yes I don't think we need to rush speccing this. If it isn't enough we can still spec App Sets later.

**Comment by cokiengchiara** [comment]
> > As we talked about this a bit on Slack, it might be interesting to include thoughts about how maybe gitops (with argo or flux) might be a good first MVP towards having basically appsets
> 
> @cokiengchiara @puja108 @piontec Biscuit are looking at gitops tooling (Flux currently more likely than Argo) for replacing draughtsman. I think we should also write docs for how customers can use gitops to manage their app CRs.
> 
> So yes I don't think we need to rush speccing this. If it isn't enough we can still spec App Sets later.

Thanks @rossf7 I think we're still achieving same goal, but we're not creating our own code, right? 

Maybe I'll rename this to: Write docs: How to use gitops tooling to achieve 'App Sets' functionality. WDYT?

**Comment by rossf7** [comment]
> Maybe I'll rename this to: Write docs: How to use gitops tooling to achieve 'App Sets' functionality. 

@cokiengchiara Yes I think we should write docs and maybe create an example repo with some of our apps to get up and running quickly.

**Comment by cokiengchiara** [comment]
Should be in either Batman or Biscuit. Talk w @giantswarm/team-biscuit

**Comment by cokiengchiara** [comment]
Call scheduled for tomorrow.

**Comment by cokiengchiara** [comment]
From @pipo02mix , but customer also chatted in Slack:
* To have a rollback mechanism in app platform for apps, similar to k8s offer for deployments (see: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-back-a-deployment)

**Comment by puja108** [comment]
There could be a use case where a user wants to roll back manually. With helm that is supported (https://helm.sh/docs/helm/helm_rollback/), but I'm not sure if a user of app-operator could trigger this manually. The thing is that with gitops tooling it the rollback would just set the version field of the app CR back to the previous version, not sure if that would result in an actual rollback, or in a rollout with an old version. Additionally, if we support this use case we need to test downgrades.
Also imagine a use case where the error is not found in normal testing but later in usage and the rollback needs to be triggered manually.

**Comment by rossf7** [comment]
> With helm that is supported (https://helm.sh/docs/helm/helm_rollback/), but I'm not sure if a user of app-operator could trigger this manually. 

@puja108 We support helm rollback in chart-operator if the `chart-operator.giantswarm.io/force-helm-upgrade` annotation is true. But this isn't easily available or documented for users.

We implemented this for Helm 3 as an alternative to `helm upgrade --force` which doesn't recreate resources so isn't that useful for us. https://github.com/giantswarm/giantswarm/issues/11376

> The thing is that with gitops tooling it the rollback would just set the version field of the app CR back to the previous version, not sure if that would result in an actual rollback, or in a rollout with an old version. 

I agree with gitops rollback is easier and having rollback support in App Platform could be a useful feature. But I think it would be better to have a separate issue for that.

@cokiengchiara WDYT?

**Comment by cokiengchiara** [comment]
> I agree with gitops rollback is easier and having rollback support in App Platform could be a useful feature. But I think it would be better to have a separate issue for that.
> 
> @cokiengchiara WDYT?

Yes, please. Can you create issue and link? 🙏

**Comment by rossf7** [comment]
@cokiengchiara I created https://github.com/giantswarm/roadmap/issues/337 for rollback support and updated the use cases section to link it.

**Comment by cokiengchiara** [comment]
From customer in [slack](https://gigantic.slack.com/archives/C6L8J93N0/p1621407796012900?thread_ts=1621351896.012400&cid=C6L8J93N0):

They have plans to run helm version exporter in order to **increase transparency and even alert on drifts between current/desired version**. 

This would probably be **a game changer** for the way we manage app and a potential driver for us to migrate our helm charts to their own app catalogue (not a commitment that this will happen though :-P). For customer app catalogs, see:  https://github.com/giantswarm/roadmap/issues/29

**Comment by cokiengchiara** [comment]
@pipo02mix mentioned this in [slack](https://gigantic.slack.com/archives/C37M39TV1/p1627292193120900) today, just to emphasize this is still relevant for customer, especially the rollback support.

**Comment by puja108** [comment]
cc @JosephSalisbury as this kinda depends on the general gitops as a interface to MCs story, the use cases mentioned here are slightly different from the ones around managing clusters, but in a way related to general features the gitops interface/tool might be able to offer.

**Comment by piontec** [comment]
Obi specific ticket: https://github.com/giantswarm/obi/issues/18

**Comment by MarcelMue** [comment]
I just read through this ticket while looking for a specific issue on our board. IMO we can close this or WDYT is still missing for this to be possible @piontec @gianfranco-l ?

**Comment by gianfranco-l** [comment]
I think this can be closed but i leave the final word to @piontec

**Comment by piontec** [comment]
Yes, let's close this one and create a new list of stuff remaining

**Comment by piontec** [comment]
Yes, let's close and create the current view of what needs to be done
